File: churnoutrate.py
# 3
import pandas as pd
import numpy as np  
# Importing the dataset
empData = pd.read_csv("C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\emp_data.csv")
# Performing EDA
empData.describe()
# Graphical Representation
import matplotlib.pyplot as plt 
# BArplot
plt.bar(height = empData.Churn_out_rate, x = np.arange(1, 11, 1))
# Histogram
plt.hist(empData.Churn_out_rate) 
# Boxplot
plt.boxplot(empData.Churn_out_rate) 
plt.bar(height = empData.Salary_hike, x = np.arange(1, 11, 1))
plt.hist(empData.Salary_hike) #histogram
plt.boxplot(empData.Salary_hike) #boxplot
# Scatter plot
plt.scatter(x = empData['Salary_hike'], y = empData['Churn_out_rate'], color = 'green') 
# correlation
np.corrcoef(empData.Salary_hike, empData.Churn_out_rate) 
# Covariance
cov_output = np.cov(empData.Salary_hike, empData.Churn_out_rate)[0, 1]
cov_output
empData.cov()
import statsmodels.formula.api as smf
# Simple Linear Regression
model = smf.ols('Churn_out_rate ~ Salary_hike', data = empData).fit()
model.summary()
pred1 = model.predict(pd.DataFrame(empData['Salary_hike']))
# Regression Line
plt.scatter(empData.Salary_hike, empData.Churn_out_rate)
plt.plot(empData.Salary_hike, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res1 = empData.Churn_out_rate - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1
# Model building on Transformed Data
# Log Transformation
# x = log(Salary_hike); y = at
plt.scatter(x = np.log(empData['Salary_hike']), y = empData['Churn_out_rate'], color = 'brown')
np.corrcoef(np.log(empData.Salary_hike), empData.Churn_out_rate) #correlation
model2 = smf.ols('Churn_out_rate ~ np.log(Salary_hike)', data = empData).fit()
model2.summary()
pred2 = model2.predict(pd.DataFrame(empData['Salary_hike']))
# Regression Line
plt.scatter(np.log(empData.Salary_hike), empData.Churn_out_rate)
plt.plot(np.log(empData.Salary_hike), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res2 = empData.Churn_out_rate - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2
# Exponential transformation
# x = Salary_hike; y = log(at)
plt.scatter(x = empData['Salary_hike'], y = np.log(empData['Churn_out_rate']), color = 'orange')
np.corrcoef(empData.Salary_hike, np.log(empData.Churn_out_rate)) #correlation
model3 = smf.ols('np.log(Churn_out_rate) ~ Salary_hike', data = empData).fit()
model3.summary()
pred3 = model3.predict(pd.DataFrame(empData['Salary_hike']))
pred3_at = np.exp(pred3)
pred3_at
# Regression Line
plt.scatter(empData.Salary_hike, np.log(empData.Churn_out_rate))
plt.plot(empData.Salary_hike, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res3 = empData.Churn_out_rate - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3
# Polynomial transformation
# x = Salary_hike; x^2 = Salary_hike*Salary_hike; y = log(at)
model4 = smf.ols('np.log(Churn_out_rate) ~ Salary_hike+ I(Salary_hike*Salary_hike)', data = empData).fit()
model4.summary()
pred4 = model4.predict(pd.DataFrame(empData))
pred4_at = np.exp(pred4)
pred4_at
# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = empData.iloc[:, 0:1].values
X_poly = poly_reg.fit_transform(X)
# y = empData.iloc[:, 1].values
plt.scatter(empData.Salary_hike, np.log(empData.Churn_out_rate))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res4 = empData.Churn_out_rate - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4
# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse
# The best model
from sklearn.model_selection import train_test_split
train, test = train_test_split(empData, test_size = 0.3)
finalmodel = smf.ols('np.log(Churn_out_rate) ~ Salary_hike + I(Salary_hike*Salary_hike)', data = train).fit()
finalmodel.summary()
# Predict on test data
test_pred = finalmodel.predict(pd.DataFrame(test))
pred_test_Churn_out_rate = np.exp(test_pred)
pred_test_Churn_out_rate
# Model Evaluation on Test data
test_res = test.Churn_out_rate - pred_test_Churn_out_rate
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse
# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
pred_train_Churn_out_rate = np.exp(train_pred)
pred_train_Churn_out_rate
# Model Evaluation on train data
train_res = train.Churn_out_rate - pred_train_Churn_out_rate
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse

File: churnoutrate.R
# 3
library(readr)
# Importing the dataset
emp <- read.csv('C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\emp_data.csv', header = T)
View(emp)
# Performing EDA
summary(emp)
library(Hmisc)
describe(emp)
library(lattice) 
# Graphical exploration
# Dotplot
dotplot(emp$Salary_hike, main = "Dot Plot of Waist Circumferences")
dotplot(emp$Churn_out_rate, main = "Dot Plot of Adipose Tissue Areas")
# Boxplot
boxplot(emp$Salary_hike, col = "dodgerblue4")
boxplot(emp$Churn_out_rate, col = "red", horizontal = T)
# Histogram
hist(emp$Salary_hike)
hist(emp$Churn_out_rate)
# Normal QQ plot
qqnorm(emp$Salary_hike)
qqline(emp$Salary_hike)
qqnorm(emp$Churn_out_rate)
qqline(emp$Churn_out_rate)
hist(emp$Salary_hike, prob = TRUE)          # prob=TRUE for probabilities not counts
lines(density(emp$Salary_hike))             # add a density estimate with defaults
lines(density(emp$Salary_hike, adjust = 2), lty = "dotted")   # add another "smoother" density
hist(emp$Churn_out_rate, prob = TRUE)          # prob=TRUE for probabilities not counts
lines(density(emp$Churn_out_rate))             # add a density estimate with defaults
lines(density(emp$Churn_out_rate, adjust = 2), lty = "dotted")   # add another "smoother" density
# Bi-variate analysis
# Scatter plot
plot(emp$Salary_hike,emp$Churn_out_rate, main = "Scatter Plot", col = "Dodgerblue4", col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Salary_hike", ylab = "Churn_out_rate", pch = 20)  # plot(x,y)
# Correlation Coefficient
cor(emp$Salary_hike, emp$Churn_out_rate)
# Co-variance
cov(emp$Salary_hike, emp$Churn_out_rate)
# Linear Regression model
reg <- lm(emp$Churn_out_rate~ emp$Salary_hike, data = emp)
summary(reg)
confint(reg, level = 0.95)
pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)
View(pred)
# ggplot for adding Regression line for data
library(ggplot2)
ggplot(data = emp, aes(emp$Salary_hike, emp$Churn_out_rate) ) + geom_point() + stat_smooth(method = lm, formula = y ~ x)
# Evaluation the model for fitness 
cor(pred$fit,  emp$Churn_out_rate)
reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse
# Transformation Techniques
# input = log(x); output = y
plot(log(emp$Salary_hike), emp$Churn_out_rate)
cor(log(emp$Salary_hike), emp$Churn_out_rate)
reg_log <- lm(emp$Churn_out_rate ~ log(emp$Salary_hike), data = emp)
summary(reg_log)
confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")
pred <- as.data.frame(pred)
cor(pred$fit, emp$Churn_out_rate)
rmse <- sqrt(mean(reg_log$residuals^2))
rmse
# Regression line for data
ggplot(data = emp, aes(log(emp$Salary_hike), emp$Churn_out_rate) ) + geom_point() + stat_smooth(method = lm, formula = y ~ log(x))

# Log transformation applied on 'y'
# input = x; output = log(y)
plot(emp$Salary_hike, log(emp$Churn_out_rate))
cor(emp$Salary_hike, log(emp$Churn_out_rate))
reg_log1 <- lm(log(emp$Churn_out_rate) ~ emp$Salary_hike, data = emp)
summary(reg_log1)
predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))
# Exponential function
pred <- exp(predlog)
pred <- as.data.frame(pred)
cor(pred$fit, emp$Churn_out_rate)
res_log1 = emp$Churn_out_rate - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse
# Regression line for data
ggplot(data = emp, aes( emp$Salary_hike, log(emp$Churn_out_rate )) ) + geom_point() + stat_smooth(method = lm, formula = log(y) ~ x)
# Non-linear models = Polynomial models
# input = x & x^2 (2-degree) and output = log(y)
reg2 <- lm(log(emp$Churn_out_rate) ~ emp$Salary_hike + I(emp$Salary_hike*emp$Salary_hike), data = emp)
summary(reg2)
predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)
pred <- as.data.frame(pred)
cor(pred$fit, emp$Churn_out_rate)
res2 = emp$Churn_out_rate - pred$fit
rmse <- sqrt(mean(res2^2))
rmse
# Regression line for data
ggplot(data = emp, aes(emp$Salary_hike, log(emp$Churn_out_rate)) ) + geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))

# Data Partition
train <- emp[1:5, ]
test <- emp[6:10, ]
log(train$Churn_out_rate)
log(test$Churn_out_rate)
plot(train$Salary_hike, log(train$Churn_out_rate))
plot(test$Salary_hike, log(test$Churn_out_rate))
model <- lm(log(emp$Churn_out_rate) ~ emp$Salary_hike + I(emp$Salary_hike * emp$Salary_hike), data = train)
summary(model)
confint(model,level=0.95)
log_res <- predict(model,interval = "confidence", newdata = test)
predict_original <- exp(log_res) # converting log values to original values
predict_original <- as.data.frame(predict_original)
test_error <- test$Churn_out_rate- predict_original$fit # calculate error/residual
test_error
test_rmse <- sqrt(mean(test_error^2))
test_rmse
log_res_train <- predict(model, interval = "confidence", newdata = train)
predict_original_train <- exp(log_res_train) # converting log values to original values
predict_original_train <- as.data.frame(predict_original_train)
train_error <- train$Churn_out_rate - predict_original_train$fit # calculate error/residual
train_error
train_rmse <- sqrt(mean(train_error^2))
train_rmse

